gdal/frmts/gtiff/geotiff.cpp

```diff
@@ -15076,192 +15104,231 @@ void GTiffDataset::ScanDirectories()
     CPLStringList aosSubdatasets;
     int iDirIndex = 0;
 
+
     FlushDirectory();
-    while( !TIFFLastDirectory( m_hTIFF )
-           && (iDirIndex == 0 || TIFFReadDirectory( m_hTIFF ) != 0) )
+
+
+    do
     {
         // Only libtiff 4.0.4 can handle between 32768 and 65535 directories.
 #if !defined(SUPPORTS_MORE_THAN_32768_DIRECTORIES)
         if( iDirIndex == 32768 )
             break;
 #endif
-        toff_t nThisDir = TIFFCurrentDirOffset(m_hTIFF);
+        toff_t nTopDir = TIFFCurrentDirOffset(m_hTIFF);
         uint32 nSubType = 0;
 
-        *m_ppoActiveDSRef = nullptr; // Our directory no longer matches this ds.
-
-        ++iDirIndex;
-
-        if( !TIFFGetField(m_hTIFF, TIFFTAG_SUBFILETYPE, &nSubType) )
-            nSubType = 0;
+        if (iDirIndex++) {
+            *m_ppoActiveDSRef = nullptr; // Our directory no longer matches this ds.
+        }
 
-        /* Embedded overview of the main image */
-        if( (nSubType & FILETYPE_REDUCEDIMAGE) != 0 &&
-            (nSubType & FILETYPE_MASK) == 0 &&
-            iDirIndex != 1 &&
-            m_nOverviewCount < 30 /* to avoid DoS */ )
+        toff_t *tmpSubIFDOffsets, *subIFDOffsets = nullptr;
+        uint16 nSubIFDs = 0;
+        if (TIFFGetField(m_hTIFF, TIFFTAG_SUBIFD, &nSubIFDs, &tmpSubIFDOffsets) && iDirIndex == 1)
         {
-            GTiffDataset *poODS = new GTiffDataset();
-            poODS->ShareLockWithParentDataset(this);
-            poODS->SetStructuralMDFromParent(this);
-            poODS->m_pszFilename = CPLStrdup(m_pszFilename);
-            if( poODS->OpenOffset( VSI_TIFFOpenChild(m_hTIFF), m_ppoActiveDSRef, nThisDir, false,
-                                   eAccess ) != CE_None
-                || poODS->GetRasterCount() != GetRasterCount() )
-            {
-                delete poODS;
-            }
-            else
+            subIFDOffsets = (toff_t *)malloc(nSubIFDs * sizeof(toff_t));
```

Comment:
for consistency with rest of the code base, use CPLMalloc() (and corresponding CPFree() instead of free)
please use static_cast<toff_t*>() as well